### ISSUE TITLE
[question] Best way to extend Conan's architecture support?

### ISSUE THREAD

**Issue by jwillikers** [conan-io/conan#10540]
What is the best way to extend Conan's architecture support for having more granular architecture settings which are converted to architecture compiler flags?
This is best illustrated by an example for adding various settings relevant for an Arm microcontroller.

Here are the relevant settings added to the `settings.yml`.

```yaml
arch:
  ...
  thumbv7em:
    float-abi: ["soft", "softfp", "hard"]
    fpu: 
      vfpv4:
      fpv5:
        precision: ["single", "double"]
    processor: [None, "cortex-m4", "cortex-m7"]
```

These settings must be converted to specific compiler flags in GCC.
Take this example profile, for instance.

```ini
[settings]
os=baremetal
arch=thumbv7em
arch.float-abi=hard
arch.fpu=vfpv4
arch.processor=cortex-m4
compiler=gcc
compiler.version=9
compiler.libcxx=libstdc++11
compiler.cppstd=17
compiler.exception=None

[conf]
tools.cmake.cmaketoolchain:system_name=Generic
```

This would be converted into the following architecture flags for GCC.

```sh
-mthumb -mfloat-abi=hard -march=armv7e-m+fp -mtune=cortex-m4
```

For a CMake project, the best point of extensibility appears to be the `CMakeToolchain` generator which has a specific block for configuring the compiler architecture blocks.
At first blush, `python_requires` appears to be the best method for extending this via the `CMakeToolchain` generator, however, this requires all dependent packages to be updated when the requirement is updated which is not as convenient as the behavior of `tool_requires` providing a `user_toolchain` which can be transitively overridden across all dependencies using that same `tool_requires`.
This also requires much less boilerplate in the Conan file itself to incorporate a `python_requires` module and its easy to incorporate a  `test_package` for this as well.

To make it possible to easily support various combinations of these architecture settings, it makes sense to generate the CMake toolchain provided by the package instead of providing a pre-configured file.

The `tool_requires` method, however, only works for static CMake files and not those generated by particular settings since `tool_requires` is evaluated in the build context and not the host context.
This requires using `requires` or `test_requires` instead and manually applying the `user_toolchain` in the generate method.
`test_requires` is not applied across dependencies in the same way that `python_requires` can not be overridden in dependencies, which makes regular `requires` the best choice.
The package id is unaffected by `tool_requires`, which doesn't adequately express the effect on the package as `python_requires` or `requires` do, though there are ways of working around this.
Additionally, using the `user_toolchain` method with a fully generated CMake toolchain can lead to the consumer's toolchain generated by Conan duplicating compiler flags and potentially overriding certain options.
Finally, these methods are all limited to one particular build system, CMake, even though these architecture flags are general enough to apply to other build systems.

With all these considerations in mind, what is the best method for extending Conan's architecture support for these highly extensible architectures like Arm and RISC-V?

- [X] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

**Comment by Hopobcn** [comment]
Hi, 
I follow a similar approach as you. A small difference is that I didn't want to modify `arch` extensively (just the bare minimum to be able to adapt from upstream changes) and added a `arch_extra` setting.  This new setting is the one modelling all architectural parameters than the toolchain requires. I believe your changes might be more robust, but take a bit more mantainability effort.

So something like this:

```
arch_extra:
    rv64:
        extensions: [None, "g", "..."]
        abi: [None, "lp64", "lp64f", "lp64d", "lp64q"]
        cmodel: [None, "medlow", "medany"]
```
        
Then I package `riscv-gnu-toolchain` according to the provided parameters. This is then applied as a `build_requires` to other recipes.

As you i'm still learning what's the best way to alter `CMakeToolchain` behavior.
For instance, to use the riscv gcc cross compiler I found that I need to alter the "generic_system" : `tc.blocks["generic_system"].values["compiler"] = self.deps_env_info["riscv-gnu-toolchain"].CC` .. but i'm unsure if there's a better way.

**Comment by jwillikers** [comment]
@Hopobcn Thanks for the support! I tweak the `generic_system` block in addition to a couple others. I think you can use the environment variables `CC` for C and `CXX` for C++ to modify the compiler name. With Conan 1.45.0, the toolchain includes a section like this:

```cmake
if(NOT DEFINED ENV{CC})
set(CMAKE_C_COMPILER arm-none-eabi-gcc)
endif()
if(NOT DEFINED ENV{CXX})
set(CMAKE_CXX_COMPILER arm-none-eabi-g++)
endif()
```

You might also want be interested in issue #9962.

**Comment by Hopobcn** [comment]
Mmm, remember that this section is only present if `compiler` & `compiler_cpp` values are present in the `generic_system`.

Btw, how do you set the environment?

Tried packaging `llvm` and defining
```
        bin_path = os.path.join(self.package_folder, "bin")
        self.buildenv_info.prepend_path('CC', os.path.join(bin_path, "clang"))
        self.buildenv_info.prepend_path('CXX', os.path.join(bin_path, "clang++"))
```
in the package_info + `VirtualBuildEnv` in the generate() but doesn't seem to be working.

**Comment by jwillikers** [comment]
@Hopobcn Good point on `compiler` and `compiler_cpp`. For configuring the environment, I've found the `buildenv_info` and `runenv_info` don't work as I'd expect them to. I might not understand them fully yet, but I've had better luck setting `env_info`. Wish I could help more with that one, but I'm still trying to figure it out.

**Comment by Hopobcn** [comment]
I found to love this:

```
        tc = CMakeToolchain(self)

        buildenv = VirtualBuildEnv(self).vars()
        if "CC" in buildenv.keys():
            tc.blocks["generic_system"].values["compiler"] = buildenv.get("CC")
        if "CXX" in buildenv.keys():
            tc.blocks["generic_system"].values["compiler_cpp"] = buildenv.get("CXX")

        tc.generate()
```
I find it better than searching for a specific build dependency. But still has the downside of having to modify all consumer packages.
